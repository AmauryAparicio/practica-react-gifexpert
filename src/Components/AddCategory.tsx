import React, {
  useState,
  ChangeEvent,
  FunctionComponent,
  Dispatch,
  SetStateAction,
} from "react";

import { func } from "prop-types";

export const AddCategory: FunctionComponent<{
  setCategories: Dispatch<SetStateAction<string[]>>;
}> = ({ setCategories }): JSX.Element => {
  const [inputValue, setInputValue] = useState<string>("");

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>): void => {
    setInputValue(e.target.value);
  };

  const handleSubmit = (e: ChangeEvent<HTMLFormElement>): void => {
    e.preventDefault();

    if (inputValue.trim().length > 2) {
      setCategories((cats: Array<string>) => [inputValue, ...cats]);
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <input type="text" value={inputValue} onChange={handleInputChange} />
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: func.isRequired,
};
