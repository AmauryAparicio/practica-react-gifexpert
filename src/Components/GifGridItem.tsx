import React, { FunctionComponent } from "react";
import { iImg } from "./../Interfaces";

export const GifGridItem: FunctionComponent<iImg> = ({ id, title, url }) => {
  return (
    <div className="card animate__animated animate__bounceInLeft">
      <img src={url} alt={title} />
      <p>{title}</p>
    </div>
  );
};
