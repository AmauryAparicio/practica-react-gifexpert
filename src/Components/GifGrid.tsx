import React, { FunctionComponent } from "react";
import { iGif } from "../Interfaces";
import { useFetchGifs } from "./../Hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
import { string } from "prop-types";

export const GifGrid: FunctionComponent<{ category: string }> = ({
  category,
}) => {
  const { data: images, loading } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>
      {loading && <p>Loading...</p>}
      <div className="card-grid">
        {images.map((img: iGif) => (
          <GifGridItem key={img.id} {...img} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: string.isRequired,
};
