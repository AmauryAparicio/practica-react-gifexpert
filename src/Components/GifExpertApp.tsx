import React, { FunctionComponent, useState } from "react";
import { AddCategory } from "./AddCategory";
import { GifGrid } from "./GifGrid";

export const GifExpertApp: FunctionComponent<{
  defaultCategories: string[];
}> = ({ defaultCategories = [] }): JSX.Element => {
  const [categories, setCategories] = useState<string[]>(defaultCategories);

  return (
    <>
      <h2>GifExpertApp</h2>
      <AddCategory setCategories={setCategories} />
      <hr />

      <ol>
        {categories.map((category: string) => (
          <GifGrid key={category} category={category} />
        ))}
      </ol>
    </>
  );
};
