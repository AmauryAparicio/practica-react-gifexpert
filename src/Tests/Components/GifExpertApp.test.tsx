import { shallow } from "enzyme";
import { GifExpertApp } from "../../Components/GifExpertApp";
import React from "react";
describe("Pruebas en <GifExpertApp/>", () => {
  test("debe mostrarse correctamente", () => {
    const wrapper = shallow(
      <GifExpertApp defaultCategories={["One Punch Man"]} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("debe mostrar una lista de categorias", () => {
    const categories: string[] = ["One Punch Man", "Dragon ball"];
    const wrapper = shallow(<GifExpertApp defaultCategories={categories} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("GifGrid").length).toBe(categories.length);
  });
});
