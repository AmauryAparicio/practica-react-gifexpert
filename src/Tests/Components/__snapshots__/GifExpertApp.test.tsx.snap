// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifExpertApp/> debe mostrar una lista de categorias 1`] = `
<Fragment>
  <h2>
    GifExpertApp
  </h2>
  <AddCategory
    setCategories={[Function]}
  />
  <hr />
  <ol>
    <GifGrid
      category="One Punch Man"
      key="One Punch Man"
    />
    <GifGrid
      category="Dragon ball"
      key="Dragon ball"
    />
  </ol>
</Fragment>
`;

exports[`Pruebas en <GifExpertApp/> debe mostrarse correctamente 1`] = `
<Fragment>
  <h2>
    GifExpertApp
  </h2>
  <AddCategory
    setCategories={[Function]}
  />
  <hr />
  <ol>
    <GifGrid
      category="One Punch Man"
      key="One Punch Man"
    />
  </ol>
</Fragment>
`;
